File: java/valid_perfect_square.java
public class Solution {
    public boolean isPerfectSquare(int num) {
        long l=0, r=num;
        long mid;
        while (l<=r) {
            mid=l+(r-l)/2;
            long prod=mid*mid;
            if (prod==num) return true;
            else if (prod<num) l=mid+1;
            else r=mid-1;
        }
        return false;
    }
}

File: python/valid-perfect-square.py
class Solution:
    def isPerfectSquare(self, num):
        """
        :type num: int
        :rtype: bool
        """
        l, r=1, num
        while l<=r:
            mid=int(l+(r-l)/2)
            if mid*mid==num: return True
            elif mid*mid<num: l=mid+1
            else: r=mid-1
        
        return False
